### PR TITLE
Monorepo Refactor

### PR DESCRIPTION
**Pull request by jshcrowthe**
# Overview

This PR refactors the SDK from a single repository, with 1 NPM module (i.e. [`firebase`](https://npm.im/firebase)), to a monorepo format with many NPM modules. Most of this PR is structural but there have been some changes made to cross-module dependencies due to the newly expose modules.

As a result of this PR we will be publishing the following packages under the [`firebase` scope](https://www.npmjs.com/search?q=scope:firebase&page=1&ranking=optimal) on NPM:

- `@firebase/app`
- `@firebase/auth`
- `@firebase/database`
- `@firebase/firestore`
- `@firebase/messaging`
- `@firebase/polyfill`
- `@firebase/storage`
- `@firebase/util`

Each of these packages will be independently versioned which will lay the groundwork for more concrete cross module communication going forward.

# Component Changes

This is a **huge** change, so to help with the review, I have broken down the review by component below. @gauntface, @mikelehen, @sphippen, @bojeil-google, @schmidt-sebastian take a look at the sections below. Feel free to review more than just your component, though I don't necessarily expect that.

## Auth

<details>
     <summary>Click to see details</summary>

### All of this work can be found here: _`/packages/auth`_

- 74246ae9475d2f3f589bd539815490c09f8a3f6f - This is the commit where this work was completed. I did the following:
     - Migrated the `src/auth.build.js` file to `packages/auth/src/auth.js`
     - Added base package files (i.e. `package.json`, `.npmignore`, etc)
     - Created a minimal build script for the auth file (see `packages/auth/gulpfile.js`)
     - Updated the firebase package dependency to include `@firebase/auth` and expose it via `firebase/auth`
     - Bundle `@firebase/auth` into the main `firebase.js` binary`
     - Added the `src/auth.js` file to the `.prettierignore` to ensure that we don't accidentally format the minified source.
</details>

## Database

<details>
     <summary>Click to see details</summary>

### All of this work can be found here: _`/packages/database`_

- 93aef0ad8aceee61dcacd1575e834505531faa7d -  The individual commits for this file were squashed to reduce merge difficulties going forward. Database, storage, messaging, and app, because of the unified build process, all followed basically the same migration path:
     - Migrated the `src/database` directory to `packages/database/src`
     - Migrated the `src/database.ts` file to `packages/database/index.ts`
     - Migrated the `tests/database` directory to `packages/database/test`
     - Refactor the paths in `packages/database/src` to properly reference `@firebase/app` and `@firebase/util` as needed.
     - Expose a `packages/database/index.node.ts` that imports the Node specific bits.

_NOTE: You can view the `database` specific changes in that commit by searching for `packages/database` in the commit above. You will see all of the file renames and required changes. Give it a minute to load the whole commit._

</details>

## Firestore

<details>
     <summary>Click to see details</summary>

### All of this work can be found here: _`/packages/firestore`_

There are several commits here that I intentionally didn't squash to help w/ the review of this. 

- 85e3b9100dc3fef7c2edf6189d6f1642ee770795 - Introduction of the firestore boilerplate
- 2430768ea0b6f3ef3ab57c19c04873bf8012e7ed - Moving of the firestore src/test files
- 7b4717fbed1d6c474d61a85cb1ef08cea3d7fadc - Remove unneeded internal typings (you can depend directly on these types as exported by the `@firebase/app` module)
- edd052baf6a77f4bee5353693339adfccd3b41d8 - Remove promise reference as this is done automatically via `@firebase/polyfill`
- e7adf9b926d278bd45307ded7cf6b20aa3ef69e3 - Refactored `firestore/index.ts` to properly reference the implementation
- d45f0b80149e6ee23aa087deee39829e5955a6f8 - Fixed test paths
- bc038c10f2c8f150f0957fb5757ec5d198aee04c - Fixed reference to the `@firebase/app`
- 63de02d554e8eddaae66f044f6a5e3d4994a8264 - Reintroduced the ambient `firestore` module typings and fixed typings references to @firebase/app
- 8302c7fd6a22f7281a2a454f4c8d00b90830f133 - Got FIrestore node tests working
- e4916e5642ae7efd8453f4156f76b93e2d52bb20 - Added firestore to the top level build and dev scripts
- bdfbeb04e0e950267481c303a1406c7bec4bfbcd - Fixed Firestore browser tests
- b519a5ec586ac7c19f1b021b90bca2fb16fa31b0 - Fixed firestore VSCode scripts

In addition `packages/webchannel-wrapper` has been imported from https://github.com/firebaseprivate/webchannel-wrapper
</details>

## Messaging

<details>
     <summary>Click to see details</summary>

### All of this work can be found here: _`/packages/messaging`_

- 93aef0ad8aceee61dcacd1575e834505531faa7d -  The individual commits for this file were squashed to reduce merge difficulties going forward. Database, storage, messaging, and app, because of the unified build process, all followed basically the same migration path:
     - Migrated the `src/messaging` directory to `packages/messaging/src`
     - Migrated the `src/messaging.ts` file to `packages/messaging/index.ts`
     - Migrated the `tests/messaging` directory to `packages/messaging/test`
     - Refactor the paths in `packages/messaging/src` to properly reference `@firebase/app` and `@firebase/util` as needed.
     - Fix `integration/messaging` to properly point to the newly created `firebase` lib

_NOTE: You can view the `messaging` specific changes in that commit by searching for `packages/messaging` in the commit above. You will see all of the file renames and required changes. Give it a minute to load the whole commit._

</details>

## Storage

<details>
     <summary>Click to see details</summary>

### All of this work can be found here: _`/packages/storage`_

- 93aef0ad8aceee61dcacd1575e834505531faa7d -  The individual commits for this file were squashed to reduce merge difficulties going forward. Database, storage, messaging, and app, because of the unified build process, all followed basically the same migration path:
     - Migrated the `src/storage` directory to `packages/storage/src`
     - Migrated the `src/storage.ts` file to `packages/storage/index.ts`
     - Migrated the `tests/storage` directory to `packages/storage/test`
     - Refactor the paths in `packages/storage/src` to properly reference `@firebase/app` and `@firebase/util` as needed.

_NOTE: You can view the `storage` specific changes in that commit by searching for `packages/storage` in the commit above. You will see all of the file renames and required changes. Give it a minute to load the whole commit._

</details>

# Other Changes

- Removed the commit message validation (no more `git cz`, or formatted git messages)
- Refactored the prettier code styling to be a `prepush` action instead of `precommit`
- Added a license header validation step to the `prepush` validation (thanks @wilhuff for this idea)
- Improved error messaging around testing/development
- Added a `test:setup` NPM script that will do most of the work of setting up a Firebase test project for the user.

# Workflow Changes

- For development, you can simply run `yarn dev` at the root of the package. This will spin up watchers on all the source files and will recompile as you develop. There is a development server available at `http://localhost:8080` that will expose the `firebase` namespace so you can play w/ your changes.

  _NOTE: I was considering also adding test watchers, but that got noisy, if we want that I can add it w/o too much trouble_

- `yarn` instead of `npm`, this new workflow requires `yarn workspaces`. I will be updating the CONTRIBUTING.md to talk about this, but I wanted to get this out. I'll update this once it is ready. In short, please use `yarn` >1.0.0 and you're good to go.

# tl;dr

Ping me w/ questions, as you have them! I'd be happy to explain anything/everything I did.

### CODE REVIEW

File: packages/firestore/src/typings/firestore.d.ts [hunk 1]
```diff
@@ -0,0 +1,1026 @@
+/**
```
Comment (mikelehen) [inline packages/firestore/src/typings/firestore.d.ts]: So it seems our type definitions are now duplicated in 3 places (here and in app's firebase.d.ts and app.d.ts). Is there any way to avoid the redundancy?  I really don't want to update 3 files in 2 different packages every time we modify an API.

Comment (jshcrowthe) [reply packages/firestore/src/typings/firestore.d.ts]: Agreed, I am looking into making it so the wrapper package (i.e. `firebase`) leverages the generated typings from each of the downstream packages. There are some limitations that make that difficult and is going to require some refactoring to make it possible. 

I opted to not refactor the typings too much initially in favor of later cleanup.

**Comment by mikelehen** [comment]
Looked over the firestore changes (thanks a ton for the pointers in the PR description! :+1: ) and everything looks good at a high level. I'll play with building / running tests / etc. a bit tomorrow and make sure everything seems to work well, but in general I think I'm a fan.

File: config/tsconfig.base.json [hunk 2]
```diff
@@ -1,24 +1,20 @@
...
       "es2015.promise",
       "dom"
     ],
-    "allowJs": true,
     "module": "commonjs",
     "moduleResolution": "node",
-    "noImplicitAny": true,
-    "outDir": "tmp",
-    "target": "ES5",
+    "noImplicitAny": false,
```
Comment (wilhuff) [inline config/tsconfig.base.json]: This is a bummer.

**Comment by wilhuff** [comment]
It looks like travis is timing out the firestore tests.

Right now task reassignment on the backend can cause an up to 10 second delay, so having a 5 second timeout per test is likely to make things noisy. We're using 20 seconds in other circumstances.

**Comment by mikelehen** [comment]
FWIW, there's a 5 second timeout, but also 3 retries... which one would hope would be enough. :-/

File: packages/firestore/package.json [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  },
+  "peerDependencies": {
+    "@firebase/app": "^0.1.0"
+  },
+  "devDependencies": {
+    "@types/chai": "^4.0.4",
+    "@types/mocha": "^2.2.42",
+    "@types/sinon": "^2.3.3",
+    "chai": "^4.1.1",
+    "gulp": "gulpjs/gulp#4.0",
```
Comment (mikelehen) [inline packages/firestore/package.json]: I think you're missing the grpc dependency.

Comment (jshcrowthe) [reply packages/firestore/package.json]: I think you're right! I'll get that added. Great catch

**Comment by mikelehen** [comment]
I'm struggling to run the Firestore tests via karma:

```
~/src/firebase-js-sdk/packages/firestore (lerna) $ `npm bin`/karma start
ts-loader: Using typescript@2.5.3 and /usr/local/google/home/mikelehen/src/firebase-js-sdk/packages/firestore/tsconfig.json
06 10 2017 09:27:56.350:INFO [karma]: Karma v1.7.1 server started at http://0.0.0.0:8080/
06 10 2017 09:27:56.352:INFO [launcher]: Launching browser ChromeHeadless with unlimited concurrency
06 10 2017 09:27:56.364:INFO [launcher]: Starting browser ChromeHeadless
06 10 2017 09:27:57.298:INFO [HeadlessChrome 0.0.0 (Linux 0.0.0)]: Connected on socket -ZKlKYOiR77i0syyAAAA with id 76998202
```

And then it just hangs (or maybe it's running the tests but not logging anything?).  Strangely if I pass `--single-run` then they actually run and stream the results as the tests run. I'm probably just using karma wrong or something...

**Comment by mikelehen** [comment]
Ahh, I guess I need `--auto-watch` or `--single-run` to get it to actually do anything.  So we need to update ~/.vscode/launch.json to do that, I guess.

File: .vscode/launch.json [hunk 4]
```diff
@@ -8,24 +8,41 @@
...
     {
       "type": "node",
       "request": "launch",
       "name": "Firestore Unit Tests (Browser)",
-      "program": "${workspaceRoot}/node_modules/gulp/bin/gulp.js",
-      "args": ["test", "--suite=firestore/unit/", "--env=browser", "--debug"]
+      "program": "${workspaceRoot}/packages/firestore/node_modules/.bin/karma",
+      "cwd": "${workspaceRoot}/packages/firestore",      
+      "args": [
+        "start",
```
Comment (mikelehen) [inline .vscode/launch.json]: I think we need --auto-watch

File: .vscode/launch.json [hunk 5]
```diff
@@ -8,24 +8,41 @@
       "type": "node",
       "request": "launch",
       "name": "Firestore Unit Tests (Node)",
-      "program": "${workspaceRoot}/node_modules/gulp/bin/gulp.js",
-      "args": ["test", "--suite=firestore/unit/", "--env=node"],
-      "port": 9229,
+      "program": "${workspaceRoot}/packages/firestore/node_modules/.bin/_mocha",
+      "cwd": "${workspaceRoot}/packages/firestore",
```
Comment (mikelehen) [inline .vscode/launch.json]: FWIW- We could probably move this whole .vscode directory under packages/firestore (each package can have its own .vscode if it wants it)...

File: packages/firestore/karma.conf.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+          { pattern: `test/unit/bootstrap.ts` },
+          { pattern: `test/integration/bootstrap.ts` }
+        ],
+    // frameworks to use
+    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
+    frameworks: ['mocha'],
+
+    client: {
+      mocha: {
+        timeout: 5000,
```
Comment (mikelehen) [inline packages/firestore/karma.conf.js]: Lets go ahead and increase this to 20000. I am seeing intermittent timeouts too, and I think 5 seconds may realistically be too slow when running against prod.

Comment (mikelehen) [reply packages/firestore/karma.conf.js]: Maybe get rid of "retries: 3" though...  I don't want to hide flaky tests, and I think this makes it weird / confusing to debug tests.

Comment (jshcrowthe) [reply packages/firestore/karma.conf.js]: Sounds good!

File: .vscode/launch.json (hunk 4 above)
Comment (jshcrowthe) [reply .vscode/launch.json]: Great point, I think this is initially disabled to help w/ startup time but for dev this is req'd

File: config/tsconfig.base.json (hunk 2 above)
Comment (jshcrowthe) [reply config/tsconfig.base.json]: Each package has its own `tsconfig.json` so you can set this for your project if you'd like.

**Comment by jshcrowthe** [comment]
Bumped the timeouts for `firestore` tests and they seem to be working. Great call @mikelehen / @wilhuff

File: packages/firestore/package.json [hunk 7]
```diff
@@ -27,6 +27,7 @@
     "@types/mocha": "^2.2.42",
     "@types/sinon": "^2.3.3",
     "chai": "^4.1.1",
+    "grpc": "^1.6.6",
```
Comment (wilhuff) [inline packages/firestore/package.json]: Just to be clear: this will accept grpc 1.7 when it comes out, yes?

Comment (jshcrowthe) [reply packages/firestore/package.json]: Yup! I always get the tilde vs. caret thing confused:

https://stackoverflow.com/questions/22343224/whats-the-difference-between-tilde-and-caret-in-package-json

> The caret, on the other hand, is more relaxed. It will update you to the most recent major version (the first number). ^1.2.3 will match any 1.x.x release including 1.3.0, but will hold off on 2.0.0.

File: .vscode/launch.json (hunk 5 above)
Comment (mikelehen) [reply .vscode/launch.json]: Sorry, that wasn't very clear.  I was thinking that we'd push people towards opening vscode inside of a specific package.  So if I'm doing firestore development, I'd open vscode in packages/firestore and then our launch configuration would be available and only our sources would show up in vscode, etc.  I think this might make sense, since this launch.json is 100% Firestore-specific anyway...  But I don't feel strongly about this. We can always do it later.

File: .vscode/launch.json [hunk 8]
```diff
@@ -8,24 +8,42 @@
...
     {
       "type": "node",
       "request": "launch",
       "name": "Firestore Integration Tests (Browser)",
-      "program": "${workspaceRoot}/node_modules/gulp/bin/gulp.js",
-      "args": ["test", "--suite=firestore/integration", "--env=browser", "--debug"]
+      "program": "${workspaceRoot}/packages/firestore/node_modules/.bin/karma",
+      "cwd": "${workspaceRoot}/packages/firestore",      
+      "args": [
+        "start",
```
Comment (mikelehen) [inline .vscode/launch.json]: you need --auto-watch here too.

Comment (jshcrowthe) [reply .vscode/launch.json]: :) Thanks for your patience

**Comment by jshcrowthe** [comment]
@mikelehen I see what you're saying.... I hadn't thought of that necessarily. I typically work w/ the whole thing open but it's because I work w/ more cross-cutting changes.

If it's alright let's come back to that. I think it's worth thinking about but I'd like to see VSCode have a little better support for having nested `.vscode` dirs so people who work w/ all context and people developing just in the package can both have a solid experience.

**Comment by bojeil-google** [comment]
Rubber stamping this.

**Comment by schmidt-sebastian** [comment]
Can you elaborate a little bit on why we had to move to 'yarn' and Node 8? I feel like the entry barrier for contributors is already pretty high and it would be nice to keep it low. The GCloud SDKs for example run and are developed with Node 4 + npm for example.

**Comment by jshcrowthe** [comment]
@schmidt-sebastian I have addressed the license issues (actually found a couple others per that comment) and so we should be good there.

`yarn` is a requirement as it is what enables the management of this type of a monorepo. W/o this type of tool, this management is a pain.

Node 8 was just to enable easier development of tooling and things within the SDK. Since we are compiling our source for use everywhere, it doesn't seem unreasonable to empower the development cycle w/ the latest syntax.
